test/start-operator.sh

```diff
@@ -9,14 +14,76 @@ source "${TEST_CONFIG:-${TEST_DIRECTORY}/test-config.sh}"
 CLUSTER=${CLUSTER:-pmem-govm}
 REPO_DIRECTORY="${REPO_DIRECTORY:-$(dirname "${TEST_DIRECTORY}")}"
 CLUSTER_DIRECTORY="${CLUSTER_DIRECTORY:-${REPO_DIRECTORY}/_work/${CLUSTER}}"
-DEPLOYMENT_DIRECTORY="${REPO_DIRECTORY}/deploy/operator"
 SSH="${CLUSTER_DIRECTORY}/ssh.0"
 KUBECTL="${SSH} kubectl" # Always use the kubectl installed in the cluster.
 
-deploy="${DEPLOYMENT_DIRECTORY}/pmem-csi-operator.yaml"
-echo "Deploying '${deploy}'..."
+function deploy_using_olm() {
+  set -x
+  echo "Starting the operator using OLM...."
+  CATALOG_DIR="${REPO_DIRECTORY}/deploy/olm-catalog"
+  BINDIR=${REPO_DIRECTORY}/_work/bin
+
+  if [ ! -d "${CATALOG_DIR}" ]; then
+    echo >&2 "'${CATALOG_DIR}' not a directory"
+    return 1
+  fi
+
+  # if there is a running operator deployment just reuse that
+  set +e
+  oupput=$(${KUBECTL} get deployment pmem-csi-operator 2>&1)
+  if [ $? -eq 0 ]; then
+    echo "Found a running operator deployment!!!"
+    exit 0
+  elif echo $oupput | grep -q -v '(NotFound)' ; then
+    echo "Failed to find the existance of an operator deployment\n   $output"
+    exit 1
+  fi
+  set -e
+
+  tmpdir="$(mktemp -d)"
+  TMP_CATALOG_DIR="${tmpdir}/olm-catalog"
+  trap 'rm -rf $tmpdir' SIGTERM SIGINT EXIT
+  # We need to alter generated catalog with custom driver/operator image
+  # So copy them to temproary location
+  cp -r ${CATALOG_DIR} ${tmpdir}/
+
+  # find the latest catalog version
+  VERSION=$(grep 'currentCSV:' ${TMP_CATALOG_DIR}/pmem-csi-operator.package.yaml | sed -r 's|.*v([0-9]+\.[0-9]+\.[0-9]+)$|\1|')
+  CSV_FILE="${TMP_CATALOG_DIR}/${VERSION}/pmem-csi-operator.clusterserviceversion.yaml"
+
+  # Update docker registry
+  # NOTE: Also updating image version to 'canary' for tests
+  if [ "${TEST_PMEM_REGISTRY}" != "" ]; then
+    sed -i -e "s^intel/pmem-csi-driver:v${VERSION}^${TEST_PMEM_REGISTRY}/pmem-csi-driver:canary^g" ${CSV_FILE}
+  fi
+  if [ "${TEST_IMAGE_PULL_POLICY}" != "" ]; then
+    sed -i -e "s^imagePullPolicy:.IfNotPresent^imagePullPolicy: ${TEST_IMAGE_PULL_POLICY}^g" ${CSV_FILE}
+  fi
+  # Patch operator deployment with appropriate label(pmem-csi.intel.com/deployment=<<deployment-name>>)
+  # OLM overwrites the deployment labels but the underneath ReplicaSet carries these labels.
```

Comment:
Is that reliable? I.e. could it happen that OLM overwrites the labels *before* the ReplicaSet is created with the labels?

Reply:
> Is that reliable? I.e. could it happen that OLM overwrites the labels _before_ the ReplicaSet is created with the labels?

The observation was that setting the deployment label once after `operator-sdk run packagemanifests` return worked fine, but that label gets lost once after the operator(pod) restart test run. My guess was that the OLM reconciles the deployment on that point and resets the labels. By this time the replica set for that deployment already created so, this should work reliably.